dspace-api/src/main/java/org/dspace/authenticate/SamlAuthentication.java

```diff
@@ -0,0 +1,750 @@
+     * found for the NetID or email and autoregister is enabled. When these conditions
+     * are met this method will create a new EPerson object.
+     *
+     * In order to create a new EPerson object there is a minimal set of metadata
+     * required: email, first name, and last name. If we don't have access to these
+     * three pieces of information then we will be unable to create a new EPerson.
+     *
+     * Note that this method only adds the minimal metadata. Any additional metadata
+     * will need to be added by the updateEPerson method.
+     *
+     * @param context The current DSpace database context
+     * @param request The current HTTP Request
+     * @return A new EPerson object or null if unable to create a new EPerson.
+     * @throws SQLException       if database error
+     * @throws AuthorizeException if authorization error
+     */
+    protected EPerson registerNewEPerson(Context context, HttpServletRequest request)
+        throws SQLException, AuthorizeException {
+
+        String nameId = findSingleAttribute(request, getNameIdAttributeName());
+
+        String emailAttributeName = getEmailAttributeName();
+        String firstNameAttributeName = getFirstNameAttributeName();
+        String lastNameAttributeName = getLastNameAttributeName();
+
+        String email = findSingleAttribute(request, emailAttributeName);
+        String firstName = findSingleAttribute(request, firstNameAttributeName);
+        String lastName = findSingleAttribute(request, lastNameAttributeName);
+
+        if (email == null || firstName == null || lastName == null) {
+            // We require that there be an email, first name, and last name.
+
+            String message = "Unable to register new eperson because we are unable to find an email address, " +
+                "first name, and last name for the user.\n";
+
+            message += "  name ID: " + nameId + "\n";
+            message += "  email: " + emailAttributeName + "=" + email + "\n";
+            message += "  first name: " + firstNameAttributeName + "=" + firstName + "\n";
+            message += "  last name: " + lastNameAttributeName + "=" + lastName;
+
+            log.error(message);
+
+            return null;
+        }
+
+        // Truncate values of parameters that are too big.
+
+        if (firstName.length() > NAME_MAX_SIZE) {
+            log.warn(
+                "Truncating eperson's first name because it is longer than " + NAME_MAX_SIZE + ": " + firstName);
+
+            firstName = firstName.substring(0, NAME_MAX_SIZE);
+        }
+        if (lastName.length() > NAME_MAX_SIZE) {
+            log.warn("Truncating eperson's last name because it is longer than " + NAME_MAX_SIZE + ": " + lastName);
+
+            lastName = lastName.substring(0, NAME_MAX_SIZE);
+        }
+
+        // Turn off authorizations to create a new user
+
+        context.turnOffAuthorisationSystem();
```

Comment:
The `context.turnOffAuthorisationSystem()` and `context.restoreAuthSystemState()` calls should be wrapped in a try / catch / finally block, so that even if an exception is thrown, we properly restore the current context auth state.

This is done properly in the `autoCreateEPersonMetadataField()` method in this class, we just need to fix up the other instances of auth bypass.

Reply:
Fixed in 8f4aacd.